### PR TITLE
PlasticHeatEnergy added to TensorMechanics

### LINKED ISSUE #8125: TensorMechanics plastic heat energy
### Description of the enhancement or error report

Material that computes stress * (plastic strain rate) , as well as derivatives of this for the Jacobian

### Rationale for the enhancement or information for reproducing the error

I need to include this type of term in the heat-flow equations in the PorousFlow module.  Plastic deformation acts as a source of heat.

### Identified impact
(i.e. Internal object changes, limited interface changes, public API change, or a list of specific applications impacted)

Is there such a thing already in TensorMechanics, @dschwen ?

### PR DESCRIPTION
**Pull request by WilkAndy**
Fixes #8125

### CODE REVIEW

**Comment by WilkAndy** [comment]
generalized_plane_strain_tm_contact.generalized_plane_strain_tm_contact  segfault when using PETSc alt.   Possibly a real bug in the test as possibly it's due to:

WARNING:  inverse_map of physical point (x,y,z)=( 1.00052, -4.68974e-06,        0) is not on element.

Pinging @hchen139 - i think this is your test, right?

**Comment by hchen139** [comment]
@WilkAndy Yes, it is. I will look into it.

@bwspenc The y component of the point should be zero due to applied Dirichlet BC. Tighten up the tolerance solves this problem ???

**Comment by permcody** [comment]
This PR passes all normal testing. PETSc-alt is now 3.7.x and I'm not sure that all of the tests are clean with new PETSc just yet. If this is approved, it can be merged ahead of time. @fdkong and others are working on 3.7.x issues.

**Comment by hchen139** [comment]
How should I check whether the updated test will pass the PETSc-alt or not? @WilkAndy

**Comment by fdkong** [comment]
I was wondering if the failed test is added in this PR. We could pass all tests except mammoth  with PETSc-3.7.x  in PR #8096. 

I would appreciate if we could resolve this issue before merge it.

**Comment by WilkAndy** [comment]
The petsc-alt test failure is not due to the material contained in this PR.   I'm happy to merge and let you guys work out the problem.

**Comment by permcody** [comment]
It sounds like we might have a regression then. I believe that @WilkAndy is correct in that this PR doesn't affect the failed test. It is likely something else that has been merged in since the last successful PETSc-alt test.

**Comment by WilkAndy** [comment]
@fdkong - when did you check the petsc-alt ?   This failing test was added 6 days ago.

**Comment by fdkong** [comment]
> @fdkong - when did you check the petsc-alt ? This failing test was added 6 days ago.

10 days ago.

File: modules/tensor_mechanics/src/kernels/PlasticHeatEnergy.C [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+
+PlasticHeatEnergy::PlasticHeatEnergy(const InputParameters & parameters) :
+    Kernel(parameters),
+    _coeff(getParam<Real>("coeff")),
+    _base_name(isParamValid("base_name") ? getParam<std::string>("base_name") + "_" : ""),
+    _plastic_heat(getMaterialProperty<Real>(_base_name + "plastic_heat")),
+    _dplastic_heat_dstrain(getMaterialProperty<RankTwoTensor>(_base_name + "dplastic_heat_dstrain")),
+    _ndisp(coupledComponents("displacements")),
+    _disp_var(3)
```
Comment (dschwen) [inline modules/tensor_mechanics/src/kernels/PlasticHeatEnergy.C]: `_disp_var(_ndisp)`

**Comment by WilkAndy** [comment]
If one of you guys is free, i'm wondering whether you could review/merge this?  In 8 hours or so i'll be on a plane which will be an ideal time to  extend this work from TensorMechanics to PorousFlow.

a

**Review by jwpeterson — APPROVED** [review]

**Comment by WilkAndy** [comment]
Thanks for the merge @jwpeterson ,

a

**Pull request merged**